{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0l690hTbJO6a"
      },
      "source": [
        "# Atelier 2 : Pr√©diction Fonctionnelle avec Tax4Fun2\n",
        "\n",
        "Cet atelier est compl√©mentaire √† l'atelier FUNGuild. Nous partons ici des fichiers de sortie de QIIME2 (donn√©es 16S) pour r√©aliser une analyse fonctionnelle des communaut√©s **procaryotes** (bact√©ries et arch√©es) √† l'aide de l'outil Tax4Fun2."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "CpMVLhTqJO6b"
      },
      "source": [
        "## Introduction\n",
        "\n",
        "Apr√®s avoir analys√© les communaut√©s fongiques avec FUNGuild, nous allons maintenant explorer les **fonctions m√©taboliques** des bact√©ries et arch√©es pr√©sentes dans nos √©chantillons de sol.\n",
        "\n",
        "**Tax4Fun2** est un outil qui pr√©dit les fonctions m√©taboliques des communaut√©s procaryotes en se basant sur leur taxonomie (s√©quences 16S rRNA). Il utilise des bases de donn√©es de g√©nomes de r√©f√©rence pour inf√©rer le contenu g√©n√©tique et les voies m√©taboliques (KEGG pathways) pr√©sentes dans la communaut√©.\n",
        "\n",
        "### Objectifs de l'atelier :\n",
        "1. Installer et configurer Tax4Fun2 dans Google Colab.\n",
        "2. Pr√©parer les donn√©es issues de QIIME2 pour Tax4Fun2.\n",
        "3. Ex√©cuter Tax4Fun2 pour pr√©dire les fonctions m√©taboliques.\n",
        "4. Visualiser les voies m√©taboliques (KEGG pathways) par √©chantillon.\n",
        "5. Corr√©ler les fonctions avec les propri√©t√©s du sol.\n",
        "\n",
        "### Diff√©rences avec FUNGuild :\n",
        "- **FUNGuild** : Guildes √©cologiques de champignons (ITS)\n",
        "- **Tax4Fun2** : Fonctions m√©taboliques de procaryotes (16S)\n",
        "- **FUNGuild** : Rapide, bas√© sur une base de donn√©es de guildes\n",
        "- **Tax4Fun2** : Plus lent, utilise des g√©nomes de r√©f√©rence pour pr√©diction"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "L5GWMnGnJO6b"
      },
      "source": [
        "## √âtape 1 : Installation de Tax4Fun2 et t√©l√©chargement des donn√©es\n",
        "\n",
        "Tax4Fun2 est un package R. Nous allons l'installer dans Google Colab avec toutes ses d√©pendances."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "FBU62Et-JO6b",
        "outputId": "b0705c7d-a92b-469f-f0b7-26fb7881e212"
      },
      "outputs": [
        {
          "metadata": {
            "tags": null
          },
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "üì¶ Installation de Tax4Fun2 et d√©pendances...\n",
            "‚è±Ô∏è  Cette √©tape peut prendre 15-20 minutes (ou plus)\n",
            "============================================================\n",
            "\n",
            "üì• Installation de R et des d√©pendances syst√®me...\n",
            "‚úÖ R install√©\n",
            "\n",
            "üì¶ Installation de rpy2 pour l'int√©gration Python-R...\n",
            "‚úÖ rpy2 install√©\n"
          ]
        },
        {
          "metadata": {
            "tags": null
          },
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Installing packages into ‚Äò/usr/local/lib/R/site-library‚Äô\n",
            "(as ‚Äòlib‚Äô is unspecified)\n",
            "\n"
          ]
        },
        {
          "metadata": {
            "tags": null
          },
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "\n",
            "üì¶ Installation des packages R requis...\n"
          ]
        },
        {
          "metadata": {
            "tags": null
          },
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: also installing the dependencies ‚Äòproxy‚Äô, ‚Äòlazyeval‚Äô, ‚Äòcheckmate‚Äô, ‚Äòzoo‚Äô, ‚Äòe1071‚Äô, ‚Äòwk‚Äô, ‚Äòrex‚Äô, ‚Äòcrosstalk‚Äô, ‚Äòxmlparsedata‚Äô, ‚Äòwhoami‚Äô, ‚Äòhunspell‚Äô, ‚ÄògridExtra‚Äô, ‚ÄòhtmlTable‚Äô, ‚Äòcolorspace‚Äô, ‚ÄòFormula‚Äô, ‚Äòmvtnorm‚Äô, ‚ÄòTH.data‚Äô, ‚Äòsandwich‚Äô, ‚ÄòSparseM‚Äô, ‚ÄòMatrixModels‚Äô, ‚ÄòclassInt‚Äô, ‚Äòs2‚Äô, ‚Äòunits‚Äô, ‚Äòprofmem‚Äô, ‚Äòplogr‚Äô, ‚Äòpixmap‚Äô, ‚Äòsp‚Äô, ‚ÄòRcppArmadillo‚Äô, ‚ÄòBiocManager‚Äô, ‚Äòcovr‚Äô, ‚ÄòDT‚Äô, ‚Äòfoghorn‚Äô, ‚Äògmailr‚Äô, ‚Äòlintr‚Äô, ‚Äòmockery‚Äô, ‚Äòpingr‚Äô, ‚Äòrhub‚Äô, ‚Äòspelling‚Äô, ‚Äòtinytest‚Äô, ‚Äòinline‚Äô, ‚Äòrbenchmark‚Äô, ‚ÄòpkgKitten‚Äô, ‚Äòggplot2movies‚Äô, ‚Äòhexbin‚Äô, ‚ÄòHmisc‚Äô, ‚Äòmapproj‚Äô, ‚Äòmaps‚Äô, ‚Äòmultcomp‚Äô, ‚Äòmunsell‚Äô, ‚Äòquantreg‚Äô, ‚Äòsf‚Äô, ‚Äòsvglite‚Äô, ‚Äòvdiffr‚Äô, ‚Äòbench‚Äô, ‚ÄòLahman‚Äô, ‚Äòlobstr‚Äô, ‚Äòmicrobenchmark‚Äô, ‚Äònycflights13‚Äô, ‚ÄòRMySQL‚Äô, ‚ÄòRPostgreSQL‚Äô, ‚ÄòRSQLite‚Äô, ‚Äòrepurrrsive‚Äô, ‚Äòade4‚Äô, ‚Äòsegmented‚Äô\n",
            "\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/proxy_0.4-27.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/lazyeval_0.2.2.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/checkmate_2.3.3.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/zoo_1.8-14.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/e1071_1.7-16.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/wk_0.9.4.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/rex_1.2.1.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/crosstalk_1.2.2.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/xmlparsedata_1.0.5.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/whoami_1.3.0.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/hunspell_3.0.6.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/gridExtra_2.3.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/htmlTable_2.4.3.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/colorspace_2.1-2.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/Formula_1.2-5.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/mvtnorm_1.3-3.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/TH.data_1.1-4.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/sandwich_3.1-1.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/SparseM_1.84-2.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/MatrixModels_0.5-4.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/classInt_0.4-11.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/s2_1.1.9.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/units_1.0-0.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/profmem_0.7.0.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/plogr_0.2.0.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/pixmap_0.4-14.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/sp_2.2-0.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/RcppArmadillo_15.0.2-2.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/BiocManager_1.30.26.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/covr_3.6.4.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/DT_0.34.0.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/foghorn_1.6.1.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/gmailr_2.0.0.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/lintr_3.2.0.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/mockery_0.4.5.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/pingr_2.0.5.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/rhub_2.0.1.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/spelling_2.3.2.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/tinytest_1.4.1.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/inline_0.3.21.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/rbenchmark_1.0.0.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/pkgKitten_0.2.4.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/ggplot2movies_0.0.1.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/hexbin_1.28.5.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/Hmisc_5.2-4.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/mapproj_1.2.12.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/maps_3.4.3.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/multcomp_1.4-29.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/munsell_0.5.1.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/quantreg_6.1.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/sf_1.0-21.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/svglite_2.2.2.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/vdiffr_1.0.8.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/bench_1.1.4.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/Lahman_13.0-0.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/lobstr_1.1.2.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/microbenchmark_1.5.0.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/nycflights13_1.0.2.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/RMySQL_0.11.1.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/RPostgreSQL_0.7-8.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/RSQLite_2.4.3.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/repurrrsive_1.1.0.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/ade4_1.7-23.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/segmented_2.1-4.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/devtools_2.4.6.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/Rcpp_1.1.0.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/ggplot2_4.0.0.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/dplyr_1.1.4.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/tidyr_1.3.1.tar.gz'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/seqinr_4.2-36.tar.gz'\n",
            "\n"
          ]
        }
      ],
      "source": [
        "print('üì¶ Installation de Tax4Fun2 et d√©pendances...')\n",
        "print('‚è±Ô∏è  Cette √©tape peut prendre 15-20 minutes (ou plus)')\n",
        "print('='*60)\n",
        "\n",
        "# Instalar R e depend√™ncias do sistema\n",
        "print('\\nüì• Installation de R et des d√©pendances syst√®me...')\n",
        "!apt-get update -qq > /dev/null 2>&1\n",
        "!apt-get install -y r-base r-base-dev libcurl4-openssl-dev libssl-dev libxml2-dev libgit2-dev > /dev/null 2>&1\n",
        "\n",
        "print('‚úÖ R install√©')\n",
        "\n",
        "# Instalar rpy2 para integra√ß√£o Python-R\n",
        "print('\\nüì¶ Installation de rpy2 pour l\\'int√©gration Python-R...')\n",
        "!pip install -q rpy2\n",
        "\n",
        "print('‚úÖ rpy2 install√©')\n",
        "\n",
        "# Carregar rpy2\n",
        "import rpy2.robjects as robjects\n",
        "from rpy2.robjects import r\n",
        "\n",
        "print('\\nüì¶ Installation des packages R requis...')\n",
        "\n",
        "# Instalar packages R\n",
        "r('''\n",
        "options(repos = c(CRAN = \"https://cloud.r-project.org/\"))\n",
        "install.packages(c(\"devtools\", \"Rcpp\", \"ggplot2\", \"dplyr\", \"tidyr\", \"seqinr\", \"biomformat\"),\n",
        "                 quiet = TRUE, dependencies = TRUE)\n",
        "''')\n",
        "\n",
        "print('‚úÖ Packages R de base install√©s')\n",
        "\n",
        "# Instalar Tax4Fun2\n",
        "print('\\nüì¶ Installation de Tax4Fun2...')\n",
        "r('''\n",
        "library(devtools)\n",
        "install_github(\"fjossandon/Tax4Fun2\", quiet = TRUE)\n",
        "''')\n",
        "\n",
        "# Verificar instala√ß√£o\n",
        "r('''\n",
        "if (require(\"Tax4Fun2\", quietly = TRUE)) {\n",
        "    print(\"‚úÖ Tax4Fun2 charg√© avec succ√®s!\")\n",
        "} else {\n",
        "    stop(\"‚ùå ERREUR : Tax4Fun2 n'a pas pu √™tre charg√©\")\n",
        "}\n",
        "''')\n",
        "\n",
        "print('\\n‚úÖ Installation termin√©e avec succ√®s!')\n",
        "\n",
        "# Clonage du d√©p√¥t GitHub\n",
        "print('\\nüì• Clonage du d√©p√¥t de donn√©es...')\n",
        "!git clone https://github.com/thgum/AtelierSYMBIOSOL.git\n",
        "\n",
        "print('\\nüìÇ Fichiers de donn√©es disponibles :')\n",
        "!ls -lh AtelierSYMBIOSOL/data/data_Tax4Fun2/"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "4uviwmZPJO6b"
      },
      "outputs": [],
      "source": [
        "# Clonage du d√©p√¥t GitHub pour acc√©der aux donn√©es\n",
        "print('üì• Clonage du d√©p√¥t de donn√©es...')\n",
        "!git clone https://github.com/thgum/AtelierSYMBIOSOL.git\n",
        "\n",
        "# Lister les fichiers pour v√©rifier\n",
        "print('\\nüìÇ Fichiers de donn√©es disponibles pour Tax4Fun2 :')\n",
        "!ls -lh AtelierSYMBIOSOL/data/data_Tax4Fun2/"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "RWp6jZYuJO6c"
      },
      "source": [
        "## √âtape 2 : Pr√©paration des fichiers d'entr√©e\n",
        "\n",
        "Tax4Fun2 n√©cessite deux fichiers principaux :\n",
        "1. **Fichier FASTA** : S√©quences repr√©sentatives des ASVs/OTUs\n",
        "2. **Table d'abondance** : Abondances des ASVs par √©chantillon\n",
        "\n",
        "Nous avons ces fichiers depuis QIIME2. V√©rifions leur format."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 886
        },
        "id": "IFwpBQ01JO6c",
        "outputId": "c833edd0-4bf4-40d9-c189-402d5f762634"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "üìä V√©rification des fichiers d'entr√©e...\n",
            "============================================================\n",
            "\n",
            "‚úÖ Fichier FASTA : 13466 s√©quences\n",
            "   Chemin : AtelierSYMBIOSOL/data/data_Tax4Fun2/dna-sequences.fasta\n",
            "\n",
            "‚úÖ Table d'OTUs : 13466 ASVs √ó 76 colonnes\n",
            "   Nombre d'√©chantillons : 75\n",
            "\n",
            "üìã Aper√ßu de la table d'OTUs :\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "                            #OTU ID  LEM59-01  LEM59-02  LEM59-03  LEM59-04  \\\n",
              "0  8497e8776933fb43f86a84ad1358fd91    2081.0    1912.0     118.0      97.0   \n",
              "1  e6fa43883a30a05c1bf46fd28fa2067d    1687.0    1618.0     242.0     157.0   \n",
              "2  7b08de2adb95b866362152360da56d53      64.0       0.0      26.0      40.0   \n",
              "3  1a1e74c6c7cacc00f00e83cb282363aa       0.0      87.0       0.0       0.0   \n",
              "4  0ca1ef9b5ffcd599c7d47f8bb65e5784     680.0     571.0     780.0     929.0   \n",
              "\n",
              "   LEM59-05  LEM59-06  LEM59-07  LEM59-08  LEM59-09  ...  LEM59-66  LEM59-67  \\\n",
              "0    1294.0      41.0      17.0     198.0    1458.0  ...    1730.0     521.0   \n",
              "1    1056.0      71.0     876.0    7187.0    9381.0  ...   10236.0    6174.0   \n",
              "2     136.0      33.0      47.0       0.0       0.0  ...    5658.0    8485.0   \n",
              "3       0.0       0.0       0.0       0.0       0.0  ...    2711.0    1859.0   \n",
              "4    1118.0     719.0     353.0     384.0     512.0  ...     726.0    1712.0   \n",
              "\n",
              "   LEM59-68  LEM59-69  LEM59-70  LEM59-71  LEM59-72  LEM59-73  LEM59-74  \\\n",
              "0    3487.0    5914.0    1542.0    8276.0     724.0    4612.0    4106.0   \n",
              "1    9376.0    7158.0    3052.0    7052.0    2458.0    2212.0    2047.0   \n",
              "2    4356.0    1776.0    8500.0    6981.0    4268.0    4352.0    4001.0   \n",
              "3     463.0     326.0    2051.0    1762.0    4152.0     927.0     318.0   \n",
              "4     512.0     404.0    1243.0     797.0     615.0     939.0     945.0   \n",
              "\n",
              "                                            taxonomy  \n",
              "0  k__Bacteria; p__Actinobacteria; c__Actinobacte...  \n",
              "1  k__Bacteria; p__Actinobacteria; c__Actinobacte...  \n",
              "2  k__Bacteria; p__Proteobacteria; c__Betaproteob...  \n",
              "3  k__Bacteria; p__Proteobacteria; c__Betaproteob...  \n",
              "4  k__Bacteria; p__Proteobacteria; c__Alphaproteo...  \n",
              "\n",
              "[5 rows x 76 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-d594a45f-d4ed-4315-baff-8fad8aa81cac\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>#OTU ID</th>\n",
              "      <th>LEM59-01</th>\n",
              "      <th>LEM59-02</th>\n",
              "      <th>LEM59-03</th>\n",
              "      <th>LEM59-04</th>\n",
              "      <th>LEM59-05</th>\n",
              "      <th>LEM59-06</th>\n",
              "      <th>LEM59-07</th>\n",
              "      <th>LEM59-08</th>\n",
              "      <th>LEM59-09</th>\n",
              "      <th>...</th>\n",
              "      <th>LEM59-66</th>\n",
              "      <th>LEM59-67</th>\n",
              "      <th>LEM59-68</th>\n",
              "      <th>LEM59-69</th>\n",
              "      <th>LEM59-70</th>\n",
              "      <th>LEM59-71</th>\n",
              "      <th>LEM59-72</th>\n",
              "      <th>LEM59-73</th>\n",
              "      <th>LEM59-74</th>\n",
              "      <th>taxonomy</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>8497e8776933fb43f86a84ad1358fd91</td>\n",
              "      <td>2081.0</td>\n",
              "      <td>1912.0</td>\n",
              "      <td>118.0</td>\n",
              "      <td>97.0</td>\n",
              "      <td>1294.0</td>\n",
              "      <td>41.0</td>\n",
              "      <td>17.0</td>\n",
              "      <td>198.0</td>\n",
              "      <td>1458.0</td>\n",
              "      <td>...</td>\n",
              "      <td>1730.0</td>\n",
              "      <td>521.0</td>\n",
              "      <td>3487.0</td>\n",
              "      <td>5914.0</td>\n",
              "      <td>1542.0</td>\n",
              "      <td>8276.0</td>\n",
              "      <td>724.0</td>\n",
              "      <td>4612.0</td>\n",
              "      <td>4106.0</td>\n",
              "      <td>k__Bacteria; p__Actinobacteria; c__Actinobacte...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>e6fa43883a30a05c1bf46fd28fa2067d</td>\n",
              "      <td>1687.0</td>\n",
              "      <td>1618.0</td>\n",
              "      <td>242.0</td>\n",
              "      <td>157.0</td>\n",
              "      <td>1056.0</td>\n",
              "      <td>71.0</td>\n",
              "      <td>876.0</td>\n",
              "      <td>7187.0</td>\n",
              "      <td>9381.0</td>\n",
              "      <td>...</td>\n",
              "      <td>10236.0</td>\n",
              "      <td>6174.0</td>\n",
              "      <td>9376.0</td>\n",
              "      <td>7158.0</td>\n",
              "      <td>3052.0</td>\n",
              "      <td>7052.0</td>\n",
              "      <td>2458.0</td>\n",
              "      <td>2212.0</td>\n",
              "      <td>2047.0</td>\n",
              "      <td>k__Bacteria; p__Actinobacteria; c__Actinobacte...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>7b08de2adb95b866362152360da56d53</td>\n",
              "      <td>64.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>26.0</td>\n",
              "      <td>40.0</td>\n",
              "      <td>136.0</td>\n",
              "      <td>33.0</td>\n",
              "      <td>47.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>5658.0</td>\n",
              "      <td>8485.0</td>\n",
              "      <td>4356.0</td>\n",
              "      <td>1776.0</td>\n",
              "      <td>8500.0</td>\n",
              "      <td>6981.0</td>\n",
              "      <td>4268.0</td>\n",
              "      <td>4352.0</td>\n",
              "      <td>4001.0</td>\n",
              "      <td>k__Bacteria; p__Proteobacteria; c__Betaproteob...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>1a1e74c6c7cacc00f00e83cb282363aa</td>\n",
              "      <td>0.0</td>\n",
              "      <td>87.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>2711.0</td>\n",
              "      <td>1859.0</td>\n",
              "      <td>463.0</td>\n",
              "      <td>326.0</td>\n",
              "      <td>2051.0</td>\n",
              "      <td>1762.0</td>\n",
              "      <td>4152.0</td>\n",
              "      <td>927.0</td>\n",
              "      <td>318.0</td>\n",
              "      <td>k__Bacteria; p__Proteobacteria; c__Betaproteob...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>0ca1ef9b5ffcd599c7d47f8bb65e5784</td>\n",
              "      <td>680.0</td>\n",
              "      <td>571.0</td>\n",
              "      <td>780.0</td>\n",
              "      <td>929.0</td>\n",
              "      <td>1118.0</td>\n",
              "      <td>719.0</td>\n",
              "      <td>353.0</td>\n",
              "      <td>384.0</td>\n",
              "      <td>512.0</td>\n",
              "      <td>...</td>\n",
              "      <td>726.0</td>\n",
              "      <td>1712.0</td>\n",
              "      <td>512.0</td>\n",
              "      <td>404.0</td>\n",
              "      <td>1243.0</td>\n",
              "      <td>797.0</td>\n",
              "      <td>615.0</td>\n",
              "      <td>939.0</td>\n",
              "      <td>945.0</td>\n",
              "      <td>k__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows √ó 76 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-d594a45f-d4ed-4315-baff-8fad8aa81cac')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-d594a45f-d4ed-4315-baff-8fad8aa81cac button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-d594a45f-d4ed-4315-baff-8fad8aa81cac');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-88af0ff7-e09d-4173-a6f2-169e94ac7d8c\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-88af0ff7-e09d-4173-a6f2-169e94ac7d8c')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-88af0ff7-e09d-4173-a6f2-169e94ac7d8c button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "‚úÖ Table de taxonomie : 13466 taxons\n",
            "\n",
            "üìã Aper√ßu de la taxonomie :\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "                         Feature ID  \\\n",
              "0  8497e8776933fb43f86a84ad1358fd91   \n",
              "1  e6fa43883a30a05c1bf46fd28fa2067d   \n",
              "2  7b08de2adb95b866362152360da56d53   \n",
              "3  1a1e74c6c7cacc00f00e83cb282363aa   \n",
              "4  0ca1ef9b5ffcd599c7d47f8bb65e5784   \n",
              "\n",
              "                                               Taxon  Confidence  \n",
              "0  k__Bacteria; p__Actinobacteria; c__Actinobacte...    0.939092  \n",
              "1  k__Bacteria; p__Actinobacteria; c__Actinobacte...    0.972020  \n",
              "2  k__Bacteria; p__Proteobacteria; c__Betaproteob...    0.898718  \n",
              "3  k__Bacteria; p__Proteobacteria; c__Betaproteob...    0.938572  \n",
              "4  k__Bacteria; p__Proteobacteria; c__Alphaproteo...    0.999556  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-ea003ed5-f52a-4c58-86dd-e7e92b9cb350\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Feature ID</th>\n",
              "      <th>Taxon</th>\n",
              "      <th>Confidence</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>8497e8776933fb43f86a84ad1358fd91</td>\n",
              "      <td>k__Bacteria; p__Actinobacteria; c__Actinobacte...</td>\n",
              "      <td>0.939092</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>e6fa43883a30a05c1bf46fd28fa2067d</td>\n",
              "      <td>k__Bacteria; p__Actinobacteria; c__Actinobacte...</td>\n",
              "      <td>0.972020</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>7b08de2adb95b866362152360da56d53</td>\n",
              "      <td>k__Bacteria; p__Proteobacteria; c__Betaproteob...</td>\n",
              "      <td>0.898718</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>1a1e74c6c7cacc00f00e83cb282363aa</td>\n",
              "      <td>k__Bacteria; p__Proteobacteria; c__Betaproteob...</td>\n",
              "      <td>0.938572</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>0ca1ef9b5ffcd599c7d47f8bb65e5784</td>\n",
              "      <td>k__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
              "      <td>0.999556</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-ea003ed5-f52a-4c58-86dd-e7e92b9cb350')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-ea003ed5-f52a-4c58-86dd-e7e92b9cb350 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-ea003ed5-f52a-4c58-86dd-e7e92b9cb350');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-6d931cd1-c198-4c8f-b0f8-f132db25094d\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-6d931cd1-c198-4c8f-b0f8-f132db25094d')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-6d931cd1-c198-4c8f-b0f8-f132db25094d button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "summary": "{\n  \"name\": \"    print(f'\\\\n\\u274c Table de taxonomie non trouv\\u00e9e : {tax_file}')\",\n  \"rows\": 5,\n  \"fields\": [\n    {\n      \"column\": \"Feature ID\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          \"e6fa43883a30a05c1bf46fd28fa2067d\",\n          \"0ca1ef9b5ffcd599c7d47f8bb65e5784\",\n          \"7b08de2adb95b866362152360da56d53\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Taxon\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 3,\n        \"samples\": [\n          \"k__Bacteria; p__Actinobacteria; c__Actinobacteria; o__Actinomycetales; f__Micrococcaceae; g__Arthrobacter\",\n          \"k__Bacteria; p__Proteobacteria; c__Betaproteobacteria; o__Burkholderiales; f__Burkholderiaceae; g__Burkholderia; s__bryophila\",\n          \"k__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rhizobiales; f__Bradyrhizobiaceae\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Confidence\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.03814200730241642,\n        \"min\": 0.8987183457732522,\n        \"max\": 0.9995560479679438,\n        \"num_unique_values\": 5,\n        \"samples\": [\n          0.9720201942986112,\n          0.9995560479679438,\n          0.8987183457732522\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {}
        }
      ],
      "source": [
        "import pandas as pd\n",
        "import os\n",
        "\n",
        "# D√©finir les chemins\n",
        "data_dir = 'AtelierSYMBIOSOL/data/data_Tax4Fun2'\n",
        "fasta_file = os.path.join(data_dir, 'dna-sequences.fasta')\n",
        "otu_file = os.path.join(data_dir, 'otu_table.txt')\n",
        "tax_file = os.path.join(data_dir, 'taxonomy.tsv')\n",
        "\n",
        "print('üìä V√©rification des fichiers d\\'entr√©e...')\n",
        "print('='*60)\n",
        "\n",
        "# V√©rifier le fichier FASTA\n",
        "if os.path.exists(fasta_file):\n",
        "    with open(fasta_file, 'r') as f:\n",
        "        fasta_lines = f.readlines()\n",
        "    n_sequences = sum(1 for line in fasta_lines if line.startswith('>'))\n",
        "    print(f'\\n‚úÖ Fichier FASTA : {n_sequences} s√©quences')\n",
        "    print(f'   Chemin : {fasta_file}')\n",
        "else:\n",
        "    print(f'\\n‚ùå Fichier FASTA non trouv√© : {fasta_file}')\n",
        "\n",
        "# Charger la table d'OTUs\n",
        "if os.path.exists(otu_file):\n",
        "    otu_table = pd.read_csv(otu_file, sep='\\t', skiprows=1)\n",
        "    print(f'\\n‚úÖ Table d\\'OTUs : {otu_table.shape[0]} ASVs √ó {otu_table.shape[1]} colonnes')\n",
        "    print(f'   Nombre d\\'√©chantillons : {otu_table.shape[1] - 1}')\n",
        "    print('\\nüìã Aper√ßu de la table d\\'OTUs :')\n",
        "    display(otu_table.head())\n",
        "else:\n",
        "    print(f'\\n‚ùå Table d\\'OTUs non trouv√©e : {otu_file}')\n",
        "\n",
        "# Charger la taxonomie\n",
        "if os.path.exists(tax_file):\n",
        "    tax_table = pd.read_csv(tax_file, sep='\\t', skiprows=1, header=None,\n",
        "                           names=['Feature ID', 'Taxon', 'Confidence'])\n",
        "    print(f'\\n‚úÖ Table de taxonomie : {tax_table.shape[0]} taxons')\n",
        "    print('\\nüìã Aper√ßu de la taxonomie :')\n",
        "    display(tax_table.head())\n",
        "else:\n",
        "    print(f'\\n‚ùå Table de taxonomie non trouv√©e : {tax_file}')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "qesD4A34JO6c"
      },
      "source": [
        "### Pr√©parer les fichiers pour Tax4Fun2\n",
        "\n",
        "Tax4Fun2 n√©cessite un format sp√©cifique. Nous allons cr√©er les fichiers n√©cessaires."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "lTzMSX3aJO6c",
        "outputId": "1eedb5a5-57b7-4263-e149-e2a17e682b62"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "‚úÖ Fichiers pr√©par√©s pour Tax4Fun2 :\n",
            "total 8.9M\n",
            "-rw-r--r-- 1 root     root 5.3M Oct 27 04:18 otu_table_for_tax4fun2.txt\n",
            "drwxr-xr-x 2 root     root 4.0K Oct 27 04:13 Ref100R\n",
            "drwxr-xr-x 2 root     root 4.0K Oct 27 02:14 results\n",
            "-rw-r--r-- 1 root     root 3.5M Oct 27 04:18 sequences.fasta\n",
            "drwxr-xr-x 8 15039045 5000 4.0K Oct 27 04:14 Tax4Fun2_ReferenceData_v2\n",
            "drwxr-xr-x 2 root     root 4.0K Oct 27 03:29 tax4fun2_results\n"
          ]
        }
      ],
      "source": [
        "# Cr√©er un r√©pertoire pour les r√©sultats\n",
        "!mkdir -p tax4fun2_output\n",
        "\n",
        "# Pr√©parer la table d'abondance (format requis par Tax4Fun2)\n",
        "# Tax4Fun2 attend : premi√®re colonne = ID, autres colonnes = √©chantillons\n",
        "otu_for_tax4fun = otu_table.copy()\n",
        "\n",
        "# Sauvegarder dans le format correct\n",
        "otu_for_tax4fun.to_csv('tax4fun2_output/otu_table_for_tax4fun2.txt',\n",
        "                       sep='\\t', index=False)\n",
        "\n",
        "# Copier le fichier FASTA\n",
        "!cp {fasta_file} tax4fun2_output/sequences.fasta\n",
        "\n",
        "print('‚úÖ Fichiers pr√©par√©s pour Tax4Fun2 :')\n",
        "!ls -lh tax4fun2_output/"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "2J3fPl60JO6c"
      },
      "source": [
        "## √âtape 3 : T√©l√©chargement de la base de donn√©es de r√©f√©rence\n",
        "\n",
        "Tax4Fun2 n√©cessite une base de donn√©es de g√©nomes de r√©f√©rence. Nous allons t√©l√©charger la base de donn√©es Ref100NR (version compacte)."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "i6OdsCVrJO6c",
        "outputId": "cfb81382-c6ff-4f08-c517-7406a37df39c"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "üì• T√©l√©chargement de la base de donn√©es Tax4Fun2 depuis Zenodo...\n",
            "‚è±Ô∏è  Fichier : 2.1 GB - Cette √©tape peut prendre 10-30 minutes\n",
            "============================================================\n"
          ]
        }
      ],
      "source": [
        "print('üì• T√©l√©chargement de la base de donn√©es Tax4Fun2 depuis Zenodo...')\n",
        "print('‚è±Ô∏è  Fichier : 2.1 GB - Cette √©tape peut prendre 10-30 minutes')\n",
        "print('='*60)\n",
        "\n",
        "# Download direto do Zenodo (link est√°vel)\n",
        "#!wget -O tax4fun2_output/Tax4Fun2_ReferenceData_v2.tar.gz \\\n",
        "#  https://zenodo.org/records/10035668/files/Tax4Fun2_ReferenceData_v2.tar.gz\n",
        "\n",
        "#print('\\nüì¶ D√©compression de la base de donn√©es...')\n",
        "#!cd tax4fun2_output && tar -xzf Tax4Fun2_ReferenceData_v2.tar.gz\n",
        "\n",
        "#print('\\nüßπ Nettoyage...')\n",
        "#!rm tax4fun2_output/Tax4Fun2_ReferenceData_v2.tar.gz\n",
        "\n",
        "#print('\\n‚úÖ Base de donn√©es t√©l√©charg√©e et extraite!')\n",
        "\n",
        "# Verificar integridade usando R\n",
        "#print('\\nüîç V√©rification de l\\'int√©grit√© de la base de donn√©es...')\n",
        "\n",
        "#import rpy2.robjects as robjects\n",
        "#r = robjects.r\n",
        "\n",
        "#r('''\n",
        "#library(Tax4Fun2)\n",
        "#tryCatch({\n",
        "#    testReferenceData(path_to_reference_data = \"tax4fun2_output/Tax4Fun2_ReferenceData_v2\")\n",
        "#    print(\"‚úÖ Base de donn√©es v√©rifi√©e et pr√™te!\")\n",
        "#}, error = function(e) {\n",
        "#    print(paste(\"‚ö†Ô∏è  V√©rification √©chou√©e:\", e$message))\n",
        " #   print(\"   (Cela peut √™tre normal, continuez avec l'analyse)\")\n",
        "#})\n",
        "#''')\n",
        "\n",
        "#print('\\nüìÇ Contenu du r√©pertoire:')\n",
        "#!ls -lh tax4fun2_output/"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "cOLqQsqwJO6c"
      },
      "source": [
        "## √âtape 4 : Ex√©cution de Tax4Fun2\n",
        "\n",
        "Maintenant nous allons ex√©cuter Tax4Fun2. Cette √©tape peut prendre **10-20 minutes** selon le nombre d'ASVs.\n",
        "\n",
        "Tax4Fun2 effectue les √©tapes suivantes :\n",
        "1. **Assignation taxonomique** des ASVs contre la base de r√©f√©rence\n",
        "2. **Pr√©diction du contenu g√©n√©tique** bas√© sur les g√©nomes de r√©f√©rence\n",
        "3. **Calcul des abondances** des voies m√©taboliques KEGG"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {
        "id": "1u35DeZsJO6c"
      },
      "outputs": [],
      "source": [
        "# --- √âtape 1: Installation des d√©pendances R et de Tax4Fun2 ---\n",
        "#print(\"üîß Installation de R et des d√©pendances...\")\n",
        "#!apt-get install -y r-base r-base-dev\n",
        "#!pip install -q rpy2==3.5.5  # Installe l'interface Python-R\n",
        "\n",
        "# Configuration de l'environnement R\n",
        "#import os\n",
        "#os.environ['R_HOME'] = '/usr/lib/R'\n",
        "#import rpy2.robjects as robjects\n",
        "#from rpy2.robjects.packages import importr\n",
        "\n",
        "# Installation de Tax4Fun2 via BiocManager\n",
        "#print(\"\\nüì¶ Installation de Tax4Fun2 (cela peut prendre quelques minutes)...\")\n",
        "#robjects.r('''\n",
        "#    if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
        "#        install.packages(\"BiocManager\")\n",
        "#    if (!requireNamespace(\"Tax4Fun2\", quietly = TRUE))\n",
        "#        BiocManager::install(\"Tax4Fun2\", update=FALSE, ask=FALSE)\n",
        "#''')\n",
        "#print(\"‚úÖ Tax4Fun2 est install√©.\")\n",
        "## --- √âtape 2: T√©l√©chargement de la base de donn√©es de r√©f√©rence ---\n",
        "#ref_db_path = \"tax4fun2_output\"\n",
        "#if not os.path.exists(ref_db_path):\n",
        "#    os.makedirs(ref_db_path)\n",
        "#    print(\"\\nüåê T√©l√©chargement de la base de donn√©es de r√©f√©rence (Fichier volumineux, ~8 Go)...\")\n",
        "#    !wget -q -P tax4fun2_output/ https://bioinfo.mpi-inf.mpg.de/projects/Tax4Fun2/Tax4Fun2_ReferenceData_v2.tar.gz\n",
        "#    print(\"\\nüì¶ D√©compression de la base de donn√©es...\")\n",
        "#    !cd tax4fun2_output && tar -xzf Tax4Fun2_ReferenceData_v2.tar.gz\n",
        "#else:\n",
        "#    print(\"\\n‚úÖ La base de donn√©es de r√©f√©rence existe d√©j√†.\")\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#!apt-get update -qq\n",
        "#!apt-get install -y ncbi-blast+"
      ],
      "metadata": {
        "id": "R8mOVuJS0Q0g"
      },
      "execution_count": 9,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!blastn -version"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "PL5z2o510w2W",
        "outputId": "3512d9d5-f95d-43cd-d886-fb81d2111738"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "blastn: 2.12.0+\n",
            " Package: blast 2.12.0, build Mar  8 2022 16:19:08\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!apt-get update -qq\n",
        "!apt-get install -y ncbi-blast+   # BLAST+ para o Tax4Fun2 (usa blastn)\n",
        "!pip install -q rpy2              # ponte Python‚ÜîR"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "syVUCA-L4_7p",
        "outputId": "a8480475-aa6c-4746-c7ad-4933d2c8cb80"
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  ncbi-data\n",
            "The following NEW packages will be installed:\n",
            "  ncbi-blast+ ncbi-data\n",
            "0 upgraded, 2 newly installed, 0 to remove and 38 not upgraded.\n",
            "Need to get 15.8 MB of archives.\n",
            "After this operation, 71.8 MB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 ncbi-data all 6.1.20170106+dfsg1-9 [3,519 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 ncbi-blast+ amd64 2.12.0+ds-3build1 [12.3 MB]\n",
            "Fetched 15.8 MB in 1s (13.2 MB/s)\n",
            "Selecting previously unselected package ncbi-data.\n",
            "(Reading database ... 126718 files and directories currently installed.)\n",
            "Preparing to unpack .../ncbi-data_6.1.20170106+dfsg1-9_all.deb ...\n",
            "Unpacking ncbi-data (6.1.20170106+dfsg1-9) ...\n",
            "Selecting previously unselected package ncbi-blast+.\n",
            "Preparing to unpack .../ncbi-blast+_2.12.0+ds-3build1_amd64.deb ...\n",
            "Unpacking ncbi-blast+ (2.12.0+ds-3build1) ...\n",
            "Setting up ncbi-data (6.1.20170106+dfsg1-9) ...\n",
            "Setting up ncbi-blast+ (2.12.0+ds-3build1) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for hicolor-icon-theme (0.17-2) ...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip show rpy2"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "o_EGpoi85-ih",
        "outputId": "ac6151cc-fdc9-45c9-b5a7-a207ffaed9f0"
      },
      "execution_count": 16,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Name: rpy2\n",
            "Version: 3.5.5\n",
            "Summary: Python interface to the R language (embedded R)\n",
            "Home-page: https://rpy2.github.io\n",
            "Author: Laurent Gautier\n",
            "Author-email: lgautier@gmail.com\n",
            "License: GPLv2+\n",
            "Location: /usr/local/lib/python3.12/dist-packages\n",
            "Requires: cffi, jinja2, pytz, tzlocal\n",
            "Required-by: \n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "qy_2iAaq6GJW",
        "outputId": "5a171c61-c439-4b17-c972-d81ebaa0efe8"
      },
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[1] 0\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "# --- CONFIGURA√á√ÉO DE CAMINHOS ---\n",
        "# Ajuste estes caminhos conforme os arquivos que voc√™ tem em /content\n",
        "otu_tsv      <- \"/content/tax4fun2_output/otu_table_for_tax4fun2.txt\"        # sua OTU table em TSV\n",
        "fasta_file   <- \"/content/tax4fun2_output/sequences.fasta\"          # seu FASTA (seqs representativas)\n",
        "ref_db_root  <- \"/content/tax4fun2_output/Tax4Fun2_ReferenceData_v2\"    # AJUSTE: onde est√° a pasta (sem ~)\n",
        "temp_folder  <- \"/content/tax4fun2_output/Ref100R\"                      # pasta de trabalho/tempor√°rios\n",
        "db_mode      <- \"Ref100NR\"\n",
        "threads      <- 2\n",
        "\n",
        "library(Tax4Fun2) #import library\n",
        "\n",
        "# 1. Run the reference blast\n",
        "runRefBlast(path_to_otus = fasta_file,\n",
        "            path_to_reference_data = ref_db_root,\n",
        "            path_to_temp_folder = temp_folder,\n",
        "            database_mode = db_mode,\n",
        "            use_force = T, num_threads = 2)\n",
        "\n",
        "# 2) makeFunctionalPrediction\n",
        "cat(\"\\nüß† makeFunctionalPrediction...\\n\")\n",
        "makeFunctionalPrediction(path_to_otu_table = otu_tsv,\n",
        "                         path_to_reference_data = ref_db_root,\n",
        "                         path_to_temp_folder = temp_folder,\n",
        "                         database_mode = db_mode,\n",
        "                         normalize_by_copy_number = TRUE,\n",
        "                         min_identity_to_reference = 0.99, normalize_pathways = TRUE)\n",
        "\n",
        "# 3) calculateFunctionalRedundancy (experimental)\n",
        "cat(\"\\nüßÆ calculateFunctionalRedundancy...\\n\")\n",
        "\n",
        "#Step 4: Calculating (multi-)functional redundancy indices (experimental)\n",
        "#calculates phylogentic distributions of KEGG functions (High FRI -> high redundancy, low FRI -> function is less redundant and #might get lost with community change)\n",
        "\n",
        "calculateFunctionalRedundancy(path_to_otu_table = otu_tsv,\n",
        "                              path_to_reference_data = ref_db_root,\n",
        "                              path_to_temp_folder = temp_folder,\n",
        "                              database_mode = db_mode,\n",
        "                              min_identity_to_reference = 0.99)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "YGthBbMWyj8l",
        "outputId": "c150175b-8894-4e2c-d377-518f02bd6e44"
      },
      "execution_count": 18,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Copy and generate database\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Reference blast started\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Reference blast finished\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Cleanup finished\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "üß† makeFunctionalPrediction...\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Using minimum idenity cutoff of 99% to nearest neighbor\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Error in FUN(X[[i]], ...) : invalid 'type' (character) of argument\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: In addition: \n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Warning messages:\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 1: \n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: In download.file(url = \"https://cloudstor.aarnet.edu.au/plus/s/EpS8jMFeLlbueyM/download\",  :\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: \n",
            " \n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:  URL 'https://cloudstor.aarnet.edu.au/plus/s/EpS8jMFeLlbueyM/download': status was 'Couldn't resolve host name'\n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 2: package ‚Äòbiomformat‚Äô is not available for this version of R\n",
            "\n",
            "A version of this package for your version of R might be available elsewhere,\n",
            "see the ideas at\n",
            "https://cran.r-project.org/doc/manuals/r-patched/R-admin.html#Installing-packages \n",
            "\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 3: \n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: In install.packages(c(\"devtools\", \"Rcpp\", \"ggplot2\", \"dplyr\", \"tidyr\",  :\n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: \n",
            " \n",
            "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:  installation of package ‚ÄòRcpp‚Äô had non-zero exit status\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Error in FUN(X[[i]], ...) : invalid 'type' (character) of argument\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "RInterpreterError",
          "evalue": "Failed to parse and evaluate line '# --- CONFIGURA√á√ÉO DE CAMINHOS ---\\n# Ajuste estes caminhos conforme os arquivos que voc√™ tem em /content\\notu_tsv      <- \"/content/tax4fun2_output/otu_table_for_tax4fun2.txt\"        # sua OTU table em TSV\\nfasta_file   <- \"/content/tax4fun2_output/sequences.fasta\"          # seu FASTA (seqs representativas)\\nref_db_root  <- \"/content/tax4fun2_output/Tax4Fun2_ReferenceData_v2\"    # AJUSTE: onde est√° a pasta (sem ~)\\ntemp_folder  <- \"/content/tax4fun2_output/Ref100R\"                      # pasta de trabalho/tempor√°rios\\ndb_mode      <- \"Ref100NR\"\\nthreads      <- 2\\n\\nlibrary(Tax4Fun2) #import library\\n\\n# 1. Run the reference blast\\nrunRefBlast(path_to_otus = fasta_file, \\n            path_to_reference_data = ref_db_root, \\n            path_to_temp_folder = temp_folder, \\n            database_mode = db_mode,\\n            use_force = T, num_threads = 2)\\n\\n# 2) makeFunctionalPrediction\\ncat(\"\\\\nüß† makeFunctionalPrediction...\\\\n\")\\nmakeFunctionalPrediction(path_to_otu_table = otu_tsv, \\n                         path_to_reference_data = ref_db_root, \\n                         path_to_temp_folder = temp_folder, \\n                         database_mode = db_mode, \\n                         normalize_by_copy_number = TRUE, \\n                         min_identity_to_reference = 0.99, normalize_pathways = TRUE)\\n\\n# 3) calculateFunctionalRedundancy (experimental)\\ncat(\"\\\\nüßÆ calculateFunctionalRedundancy...\\\\n\")\\n\\n#Step 4: Calculating (multi-)functional redundancy indices (experimental)\\n#calculates phylogentic distributions of KEGG functions (High FRI -> high redundancy, low FRI -> function is less redundant and #might get lost with community change)\\n\\ncalculateFunctionalRedundancy(path_to_otu_table = otu_tsv, \\n                              path_to_reference_data = ref_db_root, \\n                              path_to_temp_folder = temp_folder, \\n                              database_mode = db_mode, \\n                              min_identity_to_reference = 0.99)\\n'.\nR error message: \"Error in FUN(X[[i]], ...) : invalid 'type' (character) of argument\"",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mRRuntimeError\u001b[0m                             Traceback (most recent call last)",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/rpy2/ipython/rmagic.py\u001b[0m in \u001b[0;36meval\u001b[0;34m(self, code)\u001b[0m\n\u001b[1;32m    330\u001b[0m                 \u001b[0;31m# Need the newline in case the last line in code is a comment.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 331\u001b[0;31m                 \u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvisible\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mro\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"withVisible({%s\\n})\"\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0mcode\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    332\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mri\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0membedded\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mRRuntimeError\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mexception\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/rpy2/robjects/__init__.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, string)\u001b[0m\n\u001b[1;32m    458\u001b[0m         \u001b[0mp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrinterface\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstring\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 459\u001b[0;31m         \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0meval\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mp\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    460\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mconversion\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_conversion\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrpy2py\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mres\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/rpy2/robjects/functions.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    202\u001b[0m                 \u001b[0mkwargs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mr_k\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mv\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 203\u001b[0;31m         return (super(SignatureTranslatedFunction, self)\n\u001b[0m\u001b[1;32m    204\u001b[0m                 .__call__(*args, **kwargs))\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/rpy2/robjects/functions.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    125\u001b[0m                 \u001b[0mnew_kwargs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpy2rpy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 126\u001b[0;31m         \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mFunction\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__call__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mnew_args\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mnew_kwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    127\u001b[0m         \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrpy2py\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mres\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/rpy2/rinterface_lib/conversion.py\u001b[0m in \u001b[0;36m_\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     44\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 45\u001b[0;31m         \u001b[0mcdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunction\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     46\u001b[0m         \u001b[0;31m# TODO: test cdata is of the expected CType\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/rpy2/rinterface.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    814\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0merror_occured\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 815\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0membedded\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mRRuntimeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_rinterface\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_geterrmessage\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    816\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mres\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mRRuntimeError\u001b[0m: Error in FUN(X[[i]], ...) : invalid 'type' (character) of argument\n",
            "\nDuring handling of the above exception, another exception occurred:\n",
            "\u001b[0;31mRInterpreterError\u001b[0m                         Traceback (most recent call last)",
            "\u001b[0;32m/tmp/ipython-input-1891998177.py\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun_cell_magic\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'R'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'# --- CONFIGURA√á√ÉO DE CAMINHOS ---\\n# Ajuste estes caminhos conforme os arquivos que voc√™ tem em /content\\notu_tsv      <- \"/content/tax4fun2_output/otu_table_for_tax4fun2.txt\"        # sua OTU table em TSV\\nfasta_file   <- \"/content/tax4fun2_output/sequences.fasta\"          # seu FASTA (seqs representativas)\\nref_db_root  <- \"/content/tax4fun2_output/Tax4Fun2_ReferenceData_v2\"    # AJUSTE: onde est√° a pasta (sem ~)\\ntemp_folder  <- \"/content/tax4fun2_output/Ref100R\"                      # pasta de trabalho/tempor√°rios\\ndb_mode      <- \"Ref100NR\"\\nthreads      <- 2\\n\\nlibrary(Tax4Fun2) #import library\\n\\n# 1. Run the reference blast\\nrunRefBlast(path_to_otus = fasta_file, \\n            path_to_reference_data = ref_db_root, \\n            path_to_temp_folder = temp_folder, \\n            database_mode = db_mode,\\n            use_force = T, num_threads = 2)\\n\\n# 2) makeFunctionalPrediction\\ncat(\"\\\\nüß† makeFunctionalPrediction...\\\\n\")\\nmakeFunctionalPrediction(path_to_otu_table = otu_tsv, \\n                         path_to_reference_data = ref_db_root, \\n                         path_to_temp_folder = temp_folder, \\n                         database_mode = db_mode, \\n                         normalize_by_copy_number = TRUE, \\n                         min_identity_to_reference = 0.99, normalize_pathways = TRUE)...\n\u001b[0m",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/google/colab/_shell.py\u001b[0m in \u001b[0;36mrun_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m    274\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mline\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    275\u001b[0m       \u001b[0mcell\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m' '\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 276\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun_cell_magic\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmagic_name\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    277\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    278\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/IPython/core/interactiveshell.py\u001b[0m in \u001b[0;36mrun_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m   2471\u001b[0m             \u001b[0;32mwith\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbuiltin_trap\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2472\u001b[0m                 \u001b[0margs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mmagic_arg_s\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2473\u001b[0;31m                 \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2474\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2475\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m<decorator-gen-124>\u001b[0m in \u001b[0;36mR\u001b[0;34m(self, line, cell, local_ns)\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/IPython/core/magic.py\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(f, *a, **k)\u001b[0m\n\u001b[1;32m    185\u001b[0m     \u001b[0;31m# but it's overkill for just that one bit of state.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    186\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mmagic_deco\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 187\u001b[0;31m         \u001b[0mcall\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mlambda\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    188\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    189\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mcallable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/rpy2/ipython/rmagic.py\u001b[0m in \u001b[0;36mR\u001b[0;34m(self, line, cell, local_ns)\u001b[0m\n\u001b[1;32m    866\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstdout\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mendswith\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0merr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    867\u001b[0m                 \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0merr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 868\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    869\u001b[0m         \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    870\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdevice\u001b[0m \u001b[0;32min\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'png'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'svg'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/rpy2/ipython/rmagic.py\u001b[0m in \u001b[0;36mR\u001b[0;34m(self, line, cell, local_ns)\u001b[0m\n\u001b[1;32m    846\u001b[0m                     \u001b[0mreturn_output\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    847\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 848\u001b[0;31m                 \u001b[0mtext_result\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvisible\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0meval\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcode\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    849\u001b[0m                 \u001b[0mtext_output\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0mtext_result\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    850\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mvisible\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/rpy2/ipython/rmagic.py\u001b[0m in \u001b[0;36meval\u001b[0;34m(self, code)\u001b[0m\n\u001b[1;32m    333\u001b[0m                 \u001b[0;31m# Otherwise next return seems to have copy of error.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    334\u001b[0m                 \u001b[0mwarning_or_other_msg\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mflush\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 335\u001b[0;31m                 raise RInterpreterError(code, str(exception),\n\u001b[0m\u001b[1;32m    336\u001b[0m                                         warning_or_other_msg)\n\u001b[1;32m    337\u001b[0m             \u001b[0mtext_output\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mflush\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mRInterpreterError\u001b[0m: Failed to parse and evaluate line '# --- CONFIGURA√á√ÉO DE CAMINHOS ---\\n# Ajuste estes caminhos conforme os arquivos que voc√™ tem em /content\\notu_tsv      <- \"/content/tax4fun2_output/otu_table_for_tax4fun2.txt\"        # sua OTU table em TSV\\nfasta_file   <- \"/content/tax4fun2_output/sequences.fasta\"          # seu FASTA (seqs representativas)\\nref_db_root  <- \"/content/tax4fun2_output/Tax4Fun2_ReferenceData_v2\"    # AJUSTE: onde est√° a pasta (sem ~)\\ntemp_folder  <- \"/content/tax4fun2_output/Ref100R\"                      # pasta de trabalho/tempor√°rios\\ndb_mode      <- \"Ref100NR\"\\nthreads      <- 2\\n\\nlibrary(Tax4Fun2) #import library\\n\\n# 1. Run the reference blast\\nrunRefBlast(path_to_otus = fasta_file, \\n            path_to_reference_data = ref_db_root, \\n            path_to_temp_folder = temp_folder, \\n            database_mode = db_mode,\\n            use_force = T, num_threads = 2)\\n\\n# 2) makeFunctionalPrediction\\ncat(\"\\\\nüß† makeFunctionalPrediction...\\\\n\")\\nmakeFunctionalPrediction(path_to_otu_table = otu_tsv, \\n                         path_to_reference_data = ref_db_root, \\n                         path_to_temp_folder = temp_folder, \\n                         database_mode = db_mode, \\n                         normalize_by_copy_number = TRUE, \\n                         min_identity_to_reference = 0.99, normalize_pathways = TRUE)\\n\\n# 3) calculateFunctionalRedundancy (experimental)\\ncat(\"\\\\nüßÆ calculateFunctionalRedundancy...\\\\n\")\\n\\n...\nR error message: \"Error in FUN(X[[i]], ...) : invalid 'type' (character) of argument\""
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#==============================================================\n",
        "## √âTAPE 6 - Bonus: Visualisation des Voies M√©taboliques avec PathView\n",
        "#=============================================================="
      ],
      "metadata": {
        "id": "LDYgHllETGrs"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "print('üó∫Ô∏è  √âtape 6 : Visualisation des Voies M√©taboliques KEGG')\n",
        "print('='*60)\n",
        "print('\\nNous allons visualiser les voies m√©taboliques pr√©dites par Tax4Fun2')\n",
        "print('en utilisant PathView pour cr√©er des cartes m√©taboliques color√©es.')\n",
        "print('\\nVoies √† explorer:')\n",
        "print('  1. Cycle de l\\'azote (Nitrogen metabolism)')\n",
        "print('  2. Cycle du carbone (Carbon metabolism)')\n",
        "print('  3. D√©gradation des compos√©s organiques')"
      ],
      "metadata": {
        "id": "nFmCqdr7TQQ2"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "print('\\nüì¶ Installation de PathView et d√©pendances...')\n",
        "\n",
        "import rpy2.robjects as robjects\n",
        "r = robjects.r\n",
        "\n",
        "r('''\n",
        "# Installer BiocManager si n√©cessaire\n",
        "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
        "    install.packages(\"BiocManager\", quiet = TRUE)\n",
        "\n",
        "# Installer PathView et packages n√©cessaires\n",
        "BiocManager::install(c(\"pathview\", \"KEGGgraph\", \"KEGGREST\"), quiet = TRUE, update = FALSE)\n",
        "\n",
        "# Installer packages de visualisation\n",
        "install.packages(c(\"RColorBrewer\", \"viridis\"), quiet = TRUE)\n",
        "\n",
        "print(\"‚úÖ PathView et d√©pendances install√©s\")\n",
        "''')"
      ],
      "metadata": {
        "id": "ZIxJM7kQTTsf"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Pr√©paration des donn√©es par champ\n"
      ],
      "metadata": {
        "id": "yWPhwOFATgH9"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "import numpy as np\n",
        "\n",
        "# Charger les r√©sultats Tax4Fun2\n",
        "results_file = 'tax4fun2_output/results/functional_predictions.txt'\n",
        "\n",
        "if os.path.exists(results_file):\n",
        "    df_functions = pd.read_csv(results_file, sep='\\t', index_col=0)\n",
        "    print(f'‚úÖ Donn√©es charg√©es : {df_functions.shape[0]} fonctions √ó {df_functions.shape[1]} √©chantillons')\n",
        "\n",
        "    # Charger les m√©tadonn√©es de solo para identificar os campos\n",
        "    soil_file = 'AtelierSYMBIOSOL/data/data_Tax4Fun2/soil_properties_cranberry.csv'\n",
        "    df_soil = pd.read_csv(soil_file)\n",
        "\n",
        "    # Renomear primeira coluna\n",
        "    df_soil = df_soil.rename(columns={df_soil.columns[0]: 'SampleID'})\n",
        "\n",
        "    # Harmonizar IDs (LEM59-01 ‚Üí LEM5916S-01)\n",
        "    df_soil['SampleID_16S'] = df_soil['SampleID'].str.replace('LEM59-', 'LEM5916S-', regex=False)\n",
        "\n",
        "    print(f'\\nüìã M√©tadonn√©es du sol : {df_soil.shape[0]} √©chantillons')\n",
        "    print(f'   Champs disponibles : {df_soil[\"Campo\"].unique() if \"Campo\" in df_soil.columns else \"Non d√©fini\"}')\n",
        "\n",
        "    # Agrupar amostras por campo\n",
        "    if 'Campo' in df_soil.columns:\n",
        "        # Criar dicion√°rio de amostras por campo\n",
        "        samples_by_field = {}\n",
        "        for field in df_soil['Campo'].unique():\n",
        "            field_samples = df_soil[df_soil['Campo'] == field]['SampleID_16S'].tolist()\n",
        "            # Filtrar apenas amostras que existem nos resultados\n",
        "            field_samples = [s for s in field_samples if s in df_functions.columns]\n",
        "            samples_by_field[field] = field_samples\n",
        "            print(f'   Campo {field}: {len(field_samples)} √©chantillons')\n",
        "\n",
        "        # Calcular m√©dia por campo\n",
        "        df_by_field = pd.DataFrame()\n",
        "        for field, samples in samples_by_field.items():\n",
        "            if samples:\n",
        "                df_by_field[f'Campo_{field}'] = df_functions[samples].mean(axis=1)\n",
        "\n",
        "        print(f'\\n‚úÖ Donn√©es agr√©g√©es par champ : {df_by_field.shape}')\n",
        "        display(df_by_field.head())\n",
        "    else:\n",
        "        print('\\n‚ö†Ô∏è  Colonne \"Campo\" non trouv√©e, utilisation de tous les √©chantillons')\n",
        "        df_by_field = df_functions\n",
        "else:\n",
        "    print(f'‚ùå Fichier de r√©sultats non trouv√©: {results_file}')"
      ],
      "metadata": {
        "id": "_l211rgsTfgI"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# ============================================================================\n",
        "# Visualisation avec PathView - Cycle de l'Azote\n",
        "# ============================================================================\n",
        "\n",
        "print('\\nüó∫Ô∏è  Visualisation du Cycle de l\\'Azote (KEGG pathway 00910)...')\n",
        "\n",
        "# Sauvegarder dados para R\n",
        "df_by_field.to_csv('tax4fun2_output/functions_by_field.txt', sep='\\t')\n",
        "\n",
        "r('''\n",
        "library(pathview)\n",
        "library(RColorBrewer)\n",
        "\n",
        "# Charger les donn√©es\n",
        "func_data <- read.table(\"tax4fun2_output/functions_by_field.txt\",\n",
        "                        header = TRUE, sep = \"\\\\t\", row.names = 1)\n",
        "\n",
        "print(paste(\"Donn√©es charg√©es:\", nrow(func_data), \"fonctions\"))\n",
        "\n",
        "# Cr√©er le r√©pertoire pour les r√©sultats PathView\n",
        "dir.create(\"tax4fun2_output/pathview_results\", showWarnings = FALSE)\n",
        "setwd(\"tax4fun2_output/pathview_results\")\n",
        "\n",
        "# Pathway IDs pour les cycles importants\n",
        "pathways <- list(\n",
        "    \"Nitrogen_metabolism\" = \"00910\",\n",
        "    \"Carbon_metabolism\" = \"01200\",\n",
        "    \"Methane_metabolism\" = \"00680\",\n",
        "    \"Sulfur_metabolism\" = \"00920\"\n",
        ")\n",
        "\n",
        "print(\"\\\\nüó∫Ô∏è  G√©n√©ration des cartes m√©taboliques...\")\n",
        "\n",
        "# G√©n√©rer PathView pour chaque voie\n",
        "for (pathway_name in names(pathways)) {\n",
        "    pathway_id <- pathways[[pathway_name]]\n",
        "\n",
        "    tryCatch({\n",
        "        print(paste(\"\\\\n   G√©n√©ration de:\", pathway_name, \"(\", pathway_id, \")\"))\n",
        "\n",
        "        # G√©n√©rer PathView pour chaque campo\n",
        "        for (campo in colnames(func_data)) {\n",
        "            # Extrair dados de um campo\n",
        "            campo_data <- func_data[, campo, drop = FALSE]\n",
        "\n",
        "            # Remover NAs e zeros\n",
        "            campo_data <- campo_data[!is.na(campo_data[,1]) & campo_data[,1] > 0, , drop = FALSE]\n",
        "\n",
        "            if (nrow(campo_data) > 0) {\n",
        "                # Gerar PathView\n",
        "                pathview(\n",
        "                    gene.data = campo_data,\n",
        "                    pathway.id = pathway_id,\n",
        "                    species = \"ko\",  # KEGG Orthology\n",
        "                    out.suffix = paste0(pathway_name, \"_\", campo),\n",
        "                    kegg.native = TRUE,\n",
        "                    limit = list(gene = c(0, max(campo_data, na.rm = TRUE))),\n",
        "                    low = \"white\",\n",
        "                    mid = \"yellow\",\n",
        "                    high = \"red\"\n",
        "                )\n",
        "\n",
        "                print(paste(\"      ‚úÖ\", campo, \"g√©n√©r√©\"))\n",
        "            }\n",
        "        }\n",
        "\n",
        "    }, error = function(e) {\n",
        "        print(paste(\"      ‚ö†Ô∏è  Erreur pour\", pathway_name, \":\", e$message))\n",
        "    })\n",
        "}\n",
        "\n",
        "print(\"\\\\n‚úÖ Cartes m√©taboliques g√©n√©r√©es!\")\n",
        "\n",
        "# Retourner au diret√≥rio original\n",
        "setwd(\"../..\")\n",
        "\n",
        "# Lister les fichiers g√©n√©r√©s\n",
        "list.files(\"tax4fun2_output/pathview_results\", pattern = \".png$\")\n",
        "''')\n",
        "\n",
        "print('\\nüìÇ Cartes m√©taboliques g√©n√©r√©es dans: tax4fun2_output/pathview_results/')\n",
        "!ls -lh tax4fun2_output/pathview_results/*.png"
      ],
      "metadata": {
        "id": "LnnYIgpEUcht"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# ============================================================================\n",
        "# PARTIE 4 : Affichage des cartes m√©taboliques\n",
        "# ============================================================================\n",
        "\n",
        "print('\\nüñºÔ∏è  Affichage des cartes m√©taboliques...')\n",
        "\n",
        "from IPython.display import Image, display\n",
        "import glob\n",
        "\n",
        "# Trouver toutes as imagens PNG geradas\n",
        "pathway_images = glob.glob('tax4fun2_output/pathview_results/*.png')\n",
        "\n",
        "if pathway_images:\n",
        "    print(f'\\n‚úÖ {len(pathway_images)} cartes trouv√©es')\n",
        "\n",
        "    # Organizar por tipo de pathway\n",
        "    nitrogen_imgs = [img for img in pathway_images if 'Nitrogen' in img]\n",
        "    carbon_imgs = [img for img in pathway_images if 'Carbon' in img]\n",
        "\n",
        "    if nitrogen_imgs:\n",
        "        print('\\nüîµ Cycle de l\\'Azote:')\n",
        "        for img in nitrogen_imgs:\n",
        "            print(f'\\n   {os.path.basename(img)}')\n",
        "            display(Image(filename=img, width=800))\n",
        "\n",
        "    if carbon_imgs:\n",
        "        print('\\nüü¢ M√©tabolisme du Carbone:')\n",
        "        for img in carbon_imgs:\n",
        "            print(f'\\n   {os.path.basename(img)}')\n",
        "            display(Image(filename=img, width=800))\n",
        "else:\n",
        "    print('‚ö†Ô∏è  Aucune carte g√©n√©r√©e')"
      ],
      "metadata": {
        "id": "ILo56al1Uj5O"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# ============================================================================\n",
        "# PARTIE 5 : Analyse comparative entre champs\n",
        "# ============================================================================\n",
        "\n",
        "print('\\nüìä Analyse comparative des fonctions entre champs...')\n",
        "\n",
        "import matplotlib.pyplot as plt\n",
        "import seaborn as sns\n",
        "\n",
        "# S√©lectionner fonctions li√©es au cycle de l'azote\n",
        "nitrogen_functions = [\n",
        "    'K00370',  # Nitrate reductase\n",
        "    'K00371',  # Nitrite reductase\n",
        "    'K00368',  # Nitrogenase\n",
        "    'K02305',  # Nitrite reductase (NO-forming)\n",
        "    'K00531'   # Hydroxylamine reductase\n",
        "]\n",
        "\n",
        "# Filtrar fun√ß√µes dispon√≠veis\n",
        "available_n_funcs = [f for f in nitrogen_functions if f in df_by_field.index]\n",
        "\n",
        "if available_n_funcs:\n",
        "    df_nitrogen = df_by_field.loc[available_n_funcs]\n",
        "\n",
        "    # Criar heatmap\n",
        "    plt.figure(figsize=(10, 6))\n",
        "    sns.heatmap(df_nitrogen,\n",
        "                annot=True,\n",
        "                fmt='.2f',\n",
        "                cmap='YlOrRd',\n",
        "                cbar_kws={'label': 'Abondance relative'})\n",
        "    plt.title('Fonctions du Cycle de l\\'Azote par Champ',\n",
        "              fontsize=14, fontweight='bold')\n",
        "    plt.xlabel('Champ', fontsize=12)\n",
        "    plt.ylabel('Fonction KEGG', fontsize=12)\n",
        "    plt.tight_layout()\n",
        "    plt.savefig('tax4fun2_output/nitrogen_functions_heatmap.png', dpi=300, bbox_inches='tight')\n",
        "    plt.show()\n",
        "\n",
        "    print('‚úÖ Heatmap sauvegard√©: tax4fun2_output/nitrogen_functions_heatmap.png')\n",
        "else:\n",
        "    print('‚ö†Ô∏è  Fonctions du cycle de l\\'azote non trouv√©es dans les r√©sultats')"
      ],
      "metadata": {
        "id": "sszTSLbRUlk4"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# ============================================================================\n",
        "# Visualisation moderne avec ggkegg (Alternative √† PathView)\n",
        "# ============================================================================\n",
        "\n",
        "print('\\nüé® Alternative moderne : Visualisation avec ggkegg...')\n",
        "\n",
        "r('''\n",
        "# Installer ggkegg (package moderne pour visualisation KEGG)\n",
        "if (!requireNamespace(\"remotes\", quietly = TRUE))\n",
        "    install.packages(\"remotes\", quiet = TRUE)\n",
        "\n",
        "remotes::install_github(\"noriakis/ggkegg\", quiet = TRUE)\n",
        "\n",
        "library(ggkegg)\n",
        "library(ggplot2)\n",
        "\n",
        "print(\"\\\\nüìä Cr√©ation d'une visualisation moderne du cycle de l'azote...\")\n",
        "\n",
        "tryCatch({\n",
        "    # T√©l√©charger le pathway\n",
        "    pathway <- pathway(\"ko00910\")  # Nitrogen metabolism\n",
        "\n",
        "    # Cr√©er la visualisation\n",
        "    p <- ggraph(pathway) +\n",
        "         geom_node_rect(aes(fill = type)) +\n",
        "         geom_edge_link(arrow = arrow(length = unit(1, \"mm\"))) +\n",
        "         theme_void() +\n",
        "         labs(title = \"Cycle de l'Azote - Visualisation Moderne\")\n",
        "\n",
        "    ggsave(\"tax4fun2_output/nitrogen_cycle_modern.png\",\n",
        "           plot = p, width = 12, height = 8, dpi = 300)\n",
        "\n",
        "    print(\"‚úÖ Visualisation moderne cr√©√©e!\")\n",
        "\n",
        "}, error = function(e) {\n",
        "    print(paste(\"‚ö†Ô∏è  ggkegg non disponible:\", e$message))\n",
        "    print(\"   (PathView reste la meilleure option)\")\n",
        "})\n",
        "''')"
      ],
      "metadata": {
        "id": "JG4X_RYuU0SU"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "B8OcnfPXJO6c"
      },
      "source": [
        "## Conclusion\n",
        "\n",
        "üéâ **F√©licitations !** Vous avez termin√© cet atelier.\n",
        "\n",
        "### Ce que vous avez appris :\n",
        "- Comment installer et configurer Tax4Fun2 dans Google Colab.\n",
        "- Comment pr√©parer les donn√©es QIIME2 pour Tax4Fun2.\n",
        "- Comment ex√©cuter Tax4Fun2 pour pr√©dire les fonctions m√©taboliques.\n",
        "- La diff√©rence entre l'analyse de guildes (FUNGuild) et l'analyse fonctionnelle (Tax4Fun2).\n",
        "\n",
        "### Pour aller plus loin :\n",
        "- Comparer les profils fonctionnels entre diff√©rents types d'√©chantillons.\n",
        "- Corr√©ler les fonctions m√©taboliques avec les propri√©t√©s du sol.\n",
        "- Explorer les voies m√©taboliques sp√©cifiques (ex: cycle de l'azote, du carbone)."
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.0"
    },
    "colab": {
      "provenance": [],
      "toc_visible": true
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}